### PR TITLE
Add support for GHE

### PR DESCRIPTION
**Pull request by sochotnicky**
By providing github-api-url we can enable support for Github Enterprise.
In some cases it might be available for octokit automatically but it's
safer to provide an override.

### CODE REVIEW

File: src/github-api.ts
Comment (kachick) [inline src/github-api.ts:149]:
```suggestion
  params: Parameters<typeof getCheckRunSummaries>[2],
```

Looks like missing to update here is the cause of current typecheck failures

File: action.yml
Comment (kachick) [inline action.yml:15]:
```suggestion
    default: ${{ github.graphql_url }}
```

How about this template address?

File: README.md
Comment (kachick) [inline README.md:164]:
```suggestion
  github-api-url: "https://ghe-host.example.net/api/v3"
```

I didn't know detail about acme wording 🙇 However the domain looks like showing domain sales. How about to use example domains?

File: README.md
Comment (kachick) [inline README.md:73]: The CI is stacked with queue state, but this formatting should be blocked by dprint.
Could you format with `dprint fmt`?

**Review by kachick — CHANGES_REQUESTED** [review]
Thanks for addressing around this feature!

I don't have experience to use GHE, but the logic and suggestion looks good.
I have reviewed about some CI failures and parameters. 🙏

File: action.yml
Comment (kachick) [inline action.yml:12]:
```suggestion
  github-graphql-url:
```

I think this will be better to clarify the difference of REST endpoints

File: src/github-api.ts
Comment (sochotnicky) [reply src/github-api.ts]: Thank you, applied!

File: README.md
Comment (sochotnicky) [reply README.md]: Thank you, applied!

File: action.yml
Comment (sochotnicky) [reply action.yml:12]: In this case - octokit graphql actually does "magic" to figure out graphql endpoint from rest endpoint here:
https://github.com/octokit/graphql.js/blob/main/src/graphql.ts#L67

If we pass graphql endpoint as `baseUrl` it would produce url such as "https://ghe.example.net/graphql/graphql".

So in this case we need to leave it as `github-api-url`. Or if you'd prefer `githib-rest-api-url` ?

File: action.yml
Comment (sochotnicky) [reply action.yml]: Good suggestion to use the github context. I kept using `github.api_url` for reasons explained in the other comment.

File: README.md
Comment (sochotnicky) [reply README.md]: I think the prettier plugin for `dprint` doesn't quite agree with my system so it refuses to run. I reformatted manually for now while I figure that out.

Comment (sochotnicky) [reply README.md]: Ran prettier manually with `prettier -w --single-quote --print-width 120 README.md`

There was one unrelated change which I ignored.

File: action.yml
Comment (kachick) [reply action.yml:12]: https://github.com/octokit/graphql.js/blob/9c0643d34f36ed558e55193438d7aa8b031ca43d/src/graphql.ts#L23
https://github.com/octokit/graphql.js/blob/9c0643d34f36ed558e55193438d7aa8b031ca43d/src/graphql.ts#L67-L72

```typescript
const GHES_V3_SUFFIX_REGEX = /\/api\/v3\/?$/;

  // workaround for GitHub Enterprise baseUrl set with /api/v3 suffix
  // https://github.com/octokit/auth-app.js/issues/111#issuecomment-657610451
  const baseUrl = parsedOptions.baseUrl || request.endpoint.DEFAULTS.baseUrl;
  if (GHES_V3_SUFFIX_REGEX.test(baseUrl)) {
    requestOptions.url = baseUrl.replace(GHES_V3_SUFFIX_REGEX, "/api/graphql");
  }
```

>If we pass graphql endpoint as baseUrl it would produce url such as "https://ghe.example.net/graphql/graphql".

The workaround seems to focus only on the `/api/v3` suffix. Did it actually append nested graphql suffix for your URL?

File: README.md
Comment (kachick) [reply README.md]: dprint is not a prettier plugin, it is a new replacement of prettier. Both have different rules and behaviors.

Providing nix and dockerfile in this repo https://github.com/kachick/wait-other-jobs/blob/f45d69523afdb4f239c354c4dddc4b2c185b8ea5/CONTRIBUTING.md#setup, however you can use it with another way https://dprint.dev/install/

File: action.yml
Comment (sochotnicky) [reply action.yml:12]: So we have `https://ghe.example.com/api/v3` for the api_url. When I use it I end with following error (redacted):
```
const error2 = new import_request_error.RequestError(toErrorMessage(data), status, {
  RequestError [HttpError]: Not Found
      at main/dist/index.js:18839:26
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Object.next (main/dist/index.js:21920:32)
      at async Function.paginate (main/dist/index.js:21968:26)
      at async getCheckRunSummaries (main/dist/index.js:25927:55)
      at async fetchOtherRunStatus (main/dist/index.js:26018:27)
      at async run (main/dist/index.js:26183:20) {
    status: 404,
    response: {
      url: 'https://ghe.example.com/api/graphql/graphql',
      status: 404,

```

Comment (kachick) [reply action.yml:12]: https://github.com/kachick/wait-other-jobs/pull/618#discussion_r1365895187

`github.api_url` may raise the error,  but your `${{ github.graphql_url }}` have same suffix? As I understand it, GitHub GraphQL API is v4. (I may not know GHE backgrounds 🙇‍♂️ )

File: README.md
Comment (sochotnicky) [reply README.md]: Right, I am on NixOS & have direnv flake integration. But:
```$ dprint fmt
Error resolving plugin https://plugins.dprint.dev/prettier-0.27.0.json: Error starting /home/<username>/.cache/dprint/cache/plugins/dprint-plugin-prettier/0.27.0/x86_64/dprint-plugin-prettier with args [--parent-pid 1458478 --init]. No such file or directory (os error 2)
```

This is because:
```
$ ldd /tmp/dprint-plugin-prettier
	linux-vdso.so.1 (0x00007ffe583d8000)
	libgcc_s.so.1 => /nix/store/rfckdjskd983ylf05jm9mlsw7y618hyr-xgcc-12.3.0-libgcc/lib/libgcc_s.so.1 (0x00007ff41e73e000)
	libpthread.so.0 => /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libpthread.so.0 (0x00007ff41e739000)
	libm.so.6 => /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libm.so.6 (0x00007ff41e659000)
	libdl.so.2 => /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libdl.so.2 (0x00007ff41e654000)
	libc.so.6 => /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib/libc.so.6 (0x00007ff41e46e000)
	/lib64/ld-linux-x86-64.so.2 => /nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8/lib64/ld-linux-x86-64.so.2 (0x00007ff420e6c000)
```

It's assuming `/lib64/ld-linux-x86-64.so.2` exists. Which means I need to use `patchelf` to run that plugin. But I'd need to figure out how to get that integrated into dprint to patch the plugin before it's run.

File: action.yml
Comment (sochotnicky) [reply action.yml:12]: In our GHE instance `${{ github.graphql_url }}` resolves to something like: `https://ghe.example.net/api/graphql` (which is correct endpoint for it). So the octokit graphql takes: `https://ghe.example.net/api/v3` and replaces `api/v3` with `api/graphql`. Otherwise it appends `graphql` (assuming baseURL is `api.github.com`). Which is why we need to just give octokit `${{ github.api_url }}`. It figures out graphql endoint on its own.

File: README.md
Comment (sochotnicky) [reply README.md]: Found a workaround so added a commit with (re)formatting

**Comment by sochotnicky** [comment]
Not sure if there's anything more I should be fixing here.

**Comment by kachick** [comment]
Sorry for my delayed reply :bow: 

Your implementation looks good to me 👍 , but I have stacked to determine the input naming `github-api-url` or others.
We don't currently use REST API here, so I would like to clarify that the input is used for graphql.

A fundamental problem is that I don't have an environment to test GitHub Enterprise Edition...

File: action.yml
Comment (kachick) [reply action.yml:12]: Thanks for sharing! I'm reading how some other actions are handling this ....

https://github.com/google-github-actions/release-please-action/pull/532

**Comment by sochotnicky** [comment]
I've rebased the changes on latest main (things are a bit simpler now). I'm still interested in getting this merged but if you don't have access to GHE and don't want to merge this support without it I'd prefer if you just rejected the PR so I can make other arrangements.

**Comment by kachick** [comment]
Sorry for the delay in pending the state for a long time.
And thank you for keeping the interest and updating the codebase. :pray: 

I have not forgotten this PR, but I could not determine any action or direction regarding it.

Basically, if I cannot maintain the feature, I would exclude it from my personal project.
However, this suggestion is valuable and similar URLs are found in several repositories.
Also, this does not include a huge difference.

So I'm planning to merge this PR in this or next month.
Please give me some days to take review current code again.

**Comment by sochotnicky** [comment]
> Sorry for the delay in pending the state for a long time. And thank you for keeping the interest and updating the codebase. 🙏

You are an open-source maintainer, providing tools/code for others to reuse for free.  You have no obligation to reply to anyone but I appreciate the gesture :-)

> I have not forgotten this PR, but I could not determine any action or direction regarding it.
> 
> Basically, if I cannot maintain the feature, I would exclude it from my personal project. 

That is absolutely a fair point and would be absolutely acceptable reason to reject my PR.

> However, this suggestion is valuable and similar URLs are found in several repositories. Also, this does not include a huge difference.

Yeah, I think with your refactoring the changes became much simpler 
 
> So I'm planning to merge this PR in this or next month. Please give me some days to take review current code again.

Great, no rush. For what it's worth I've been using this version on our GH enterprise instance for a while without issues. New version fixed a few issues we had with the matrix jobs etc.

File: README.md
Comment (kachick) [inline README.md:100]:
```suggestion
  github-api-url: 'https://ghe-host.example.net/api/v3'
```

I don't know the detail for acme and the domain owner, don't we need to use `example.*` for this use? https://www.rfc-editor.org/rfc/rfc2606.txt

File: src/schema.ts
Comment (kachick) [inline src/schema.ts:112]: I didn't know zod provides url validations 👍

File: action.yml
Comment (kachick) [inline action.yml:15]:
```suggestion
    default: '${{ github.api_url }}'
```

Can we use placeholder for getting default value?

File: README.md
Comment (kachick) [inline README.md:71]: At least for REST endpoint, this option name is good 👍 ref: https://github.com/actions/create-github-app-token/pull/88

---

```suggestion
| `github-api-url`                    | The Github API endpoint. Override for Github Enterprise usage. | `string` | `${{ github.api_url }}`                | `https://api.github.com`                 |
```

? (sorry for unformatting)

File: action.yml
Comment (kachick) [reply action.yml:12]: NOTE to me: The patch introduced in https://github.com/octokit/graphql.js/pull/186 is still active in latest https://github.com/octokit/graphql.js/blob/a801c2f72fc5e56d531b16d0f4c4946e687e32e6/src/graphql.ts#L67-L72

File: README.md
Comment (kachick) [reply README.md:100]: Ah, I guess resolving conflict drops old review points... https://github.com/kachick/wait-other-jobs/pull/618#discussion_r1365897262

Sorry for my lately response 🙇‍♂️

**Review by kachick — APPROVED** [review]
Thank you for all of your work of this PR! 👍 

I left some review comments. But I may merge this PR with adding some revised commit (or squashed commit) 💪

**Comment by kachick** [comment]
Released in https://github.com/kachick/wait-other-jobs/releases/tag/v3.5.0 🎉

**Pull request merged**
